### PR TITLE
Split RouteConfig for HTTP and HTTPS listeners

### PR DESCRIPTION
**Pull request by nak3**
Currently RouteConfig (`external_services`) is used by both HTTP
Listener and HTTPS Listener when TLS is enabled.
However, to add redirect support, we cannot share the RouteConfig as
HTTPListener needs Route with `RedirectAction`.

So, this patch introduces another RouteConfig
(`external_tls_services`) for the HTTPS Listener.

Part of https://github.com/knative-sandbox/net-kourier/pull/619

### CODE REVIEW

File: pkg/generator/ingress_translator.go [hunk 1]
```diff
@@ -174,6 +177,8 @@ func (translator *IngressTranslator) translateIngress(ctx context.Context, ingre
 			if len(wrs) != 0 {
 				routes = append(routes, envoy.NewRoute(
 					pathName, matchHeadersFromHTTPPath(httpPath), path, wrs, 0, httpPath.AppendHeaders, httpPath.RewriteHost))
+				tlsRoutes = append(tlsRoutes, envoy.NewRoute(
```
Comment (markusthoemmes) [inline pkg/generator/ingress_translator.go:180]: Do we really want to always do this, even if the Ingress has no TLS config?

File: pkg/generator/ingress_translator.go [hunk 2]
```diff
@@ -182,20 +185,34 @@ func (translator *IngressTranslator) translateIngress(ctx context.Context, ingre
 			return nil, nil
 		}
 
-		var virtualHost *route.VirtualHost
+		if len(ingress.Spec.TLS) != 0 {
+			tlsRoutes = append(tlsRoutes, routes...)
```
Comment (nak3) [inline pkg/generator/ingress_translator.go:189]: This line (Copy `routes` to `tlsRoutes`) will not work as https://github.com/knative-sandbox/net-kourier/pull/619/  needs the condition - routes and tlsRoutes are not same when Redirect option.

File: pkg/generator/ingress_translator.go (hunk 1 above)
Comment (nak3) [reply pkg/generator/ingress_translator.go]: We don't need it. Updated the PR to create tlsRoutes only when ingress has TLS config.
There are some extra `if` conditions so I always created tlsRoutes in the first PR.

File: pkg/generator/caches.go [hunk 3]
```diff
@@ -238,21 +246,24 @@ func generateListenersAndRouteConfigs(
...
+			ctx, externalTLSManager, kubeclient,
 		)
 		if err != nil {
 			return nil, nil, err
 		}
 		listeners = append(listeners, externalHTTPSEnvoyListener)
 	}
 
+	if len(sniMatches) > 0 || useHTTPSListenerWithOneCert() {
+		return listeners, []cachetypes.Resource{externalRouteConfig, externalTLSRouteConfig, internalRouteConfig}, nil
```
Comment (markusthoemmes) [inline pkg/generator/caches.go:265]: nit: Should we create a `resources` list which we append to in the respective branches so we avoid this branch here? envisioning this to be `return listeners, resources, nil` (and without the `if`)

**Review by markusthoemmes — APPROVED** [review]
/lgtm
/approve

/hold
For the actions to pass.

**Comment by nak3** [comment]
/hold cancel

All tests passed :tada:

**Pull request merged**